builder/builder.go

```diff
@@ -351,148 +426,390 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
-		ib, err := b.AddBundles(includes, force, false, false)
+
+	// Add all upstream bundles to the bundles
+	if allupstream {
+		clrbundledir := getUpstreamBundlesPath(b.UpstreamVer)
+		files, err := ioutil.ReadDir(clrbundledir)
 		if err != nil {
-			return bundleAddCount, err
+			return errors.Wrapf(err, "Failed to read upstream bundles dir: %s", clrbundledir)
+		}
+
+		for _, file := range files {
+			bundle := file.Name()
+			bundles = append(bundles, bundle)
+			fmt.Printf("Adding bundle %s from upstream bundles\n", bundle)
 		}
-		bundleAddCount += ib
 	}
 
-	if git && bundleAddCount > 0 {
-		// Save current dir so we can get back to it
+	// Add bundles to set
+	for _, bundle := range bundles {
+		bundleSet[bundle] = true
+	}
+
+	// Write final mix bundle list back to file
+	if err := b.writeMixBundleList(bundleSet); err != nil {
+		return err
+	}
+
+	if git {
 		fmt.Println("Adding git commit")
-		if err := helpers.Git("-C", b.Bundledir, "add", "."); err != nil {
-			return bundleAddCount, err
+		if err := helpers.Git("add", "."); err != nil {
+			return err
 		}
-		commitMsg := fmt.Sprintf("Added bundles from Clear Version %s\n\nBundles added: %v", b.Clearver, bundles)
-		if err := helpers.Git("-C", b.Bundledir, "commit", "-q", "-m", commitMsg); err != nil {
-			return bundleAddCount, err
+		commitMsg := fmt.Sprintf("Added bundles from local-bundles or upstream version %s\n\nBundles added: %v", b.UpstreamVer, bundles)
+		if err := helpers.Git("commit", "-q", "-m", commitMsg); err != nil {
+			return err
 		}
 	}
-	return bundleAddCount, nil
+	return nil
 }
 
-// InitMix will initialise a new swupd-client consumable "mix" with the given
-// based Clear Linux version and specified mix version.
-func (b *Builder) InitMix(clearver string, mixver string, all bool, upstreamurl string) error {
-	// Set up metadata
-	err := ioutil.WriteFile(b.Versiondir+"/.clearurl", []byte(upstreamurl), 0644)
+const (
+	// Using the Unicode "Box Drawing" group
+	treeNil = "    "                // "    "
+	treeBar = "\u2502   "           // "│   "
+	treeMid = "\u251C\u2500\u2500 " // "├── "
+	treeEnd = "\u2514\u2500\u2500 " // "└── "
+)
+
+func (b *Builder) buildTreePrintValue(bundle *bundle, level int, levelEnded []bool) string {
+	// Set up the value for this bundle
+	value := bundle.Name
+	if b.isLocalBundle(bundle.Filename) {
+		value += " (local)"
+	} else {
+		value += " (upstream)"
+	}
+
+	if level == 0 {
+		return value
+	}
+
+	var buff bytes.Buffer
+	// Add continuation bars if earlier levels have not ended
+	for i := 0; i < level-1; i++ {
+		if levelEnded[i] {
+			buff.WriteString(treeNil)
+		} else {
+			buff.WriteString(treeBar)
+		}
+	}
+
+	// Add a mid bar or an end bar
+	if levelEnded[level-1] {
+		buff.WriteString(treeEnd)
+	} else {
+		buff.WriteString(treeMid)
+	}
+
+	// Add the actual value
+	buff.WriteString(value)
+
+	return buff.String()
+}
+
+func (b *Builder) bundleTreePrint(set bundleSet, bundle string, level int, levelEnded []bool) {
+
```

Comment:
Random space?

Reply:
Where the heck did that come from!  Thanks for catching this.